cloudvolume/storage/storage_interfaces.py

```diff
@@ -87,13 +90,15 @@ def put_file(
   def get_file(self, file_path, start=None, end=None):
     path = self.get_path_to_file(file_path)
 
-    compressed = os.path.exists(path + '.gz')
-      
-    if compressed:
+    if os.path.exists(path + '.gz'):
+      encoding = "gzip"
       path += '.gz'
+    elif os.path.exists(path + '.br'):
```

Comment:
Is there an advantage to using the brotli decompressor vs the gzip decompressor? My understanding is that cross compatibility was supposed to be brotli's big contribution. If the difference is slight, it might make sense to just write all these files as .gz to avoid adding an additional syscall + FS access here.

Reply:
Since it is very uncommon for datasets to have mixed compression types, maybe it would be a good idea to cache the result of the first successful operation?

Reply:
They will need to have a different extension from gzip, as they need a different decompressor, and `.br` seems to be the standard.

Agreed on caching compression type.  Any preference as to how that's implemented?

Reply:
I looked into a bit more. The caching issue is tricky because not every `Storage` object is pointed at the image data. Ideally, we'd pass in a flag that enables this caching in parts of the code where such assumptions are warranted, and it would then take effect within the context of a given `get_files` call. However, this requires enough work that I'm not sure it's worth it until we experience an issue.  

Do you expect `br` to rapidly eclipse the usage of `gz`? If so, the way you've structured the if statement is ideal. Otherwise, the `.gz` test should probably go first. Based on your data below, `br` does seem advantageous.

Reply:
compressing brotli sometimes takes orders of magnitude longer.  not sure it's worth it with every dataset.  hard to say if brotli will ever eclipse gzip...

Reply:
Is that true even at compression level 6?

Reply:
I was experimenting with a uint8 dataset today and was running into some of that, yes. Should have something to share tomorrow.

Reply:
What about adding a way to specify the compression level?

Reply:
I made an effort today to add this as a parameter.  It's quite possible though that I may have missed some places, it needed to be added in a lot of different places!

Reply:
This is really cool and useful for gzip as well! :) What were your results with the uint8 data like?

Reply:
Compression rate good, decompression rate good, but for some reason compression time went way up for levels > 4.  Investigating further on another machine.

Reply:
Brotli compression is single threaded as far as I know.  I'm specifying cloudvolume w/ Parallel=False.  Still seeing lots of cores being used.

On two machines, seeing a big difference in performance w/ brotli, but seemingly only with the 8-bit data.

Below are the two different runs:

32 thread (16 cores) server:
```
Create time (s)
('br', 1): 8.20
('br', 2): 7.41
('br', 3): 7.74
('br', 4): 7.77
('br', 5): 8.04
('br', 6): 9.18
('br', 7): 13.71
(False, None): 6.61
('gzip', None): 7.59

Size (MB)
('br', 1): 444.20
('br', 2): 425.68
('br', 3): 425.15
('br', 4): 436.03
('br', 5): 349.76
('br', 6): 347.46
('br', 7): 347.62
(False, None): 733.61
('gzip', None): 392.54
```

8 threads (4 cores) laptop:
```
Create time (s)
('br', 1): 4.27
('br', 2): 4.23
('br', 3): 4.25
('br', 4): 5.60
('br', 5): 23.29
('br', 6): 43.60
('br', 7): 66.36
(False, None): 3.23
('gzip', None): 7.10

Size (MB)
('br', 1): 444.20
('br', 2): 425.68
('br', 3): 425.15
('br', 4): 436.03
('br', 5): 349.76
('br', 6): 347.46
('br', 7): 347.62
(False, None): 733.61
('gzip', None): 392.54
```

The compression ratio is at least the same across the two different machines.

Reply:
Further tests make be believe that there's nothing specific about 8 bit data that is a problem -- perhaps it's just that a larger dataset and the number of threads available that slows down the compression.

Here's some tests on the exact same 16 bit dataset, both at 16 bit and also coerced to 8 bit.  

In this case, I do not see anything peculiar about brotli with 8 bit data, and it actually performs much faster.  I think changing brotli default quality=5 might be better though.  It will be closer in performance to gzip I think, but still give better compression.

`16bit`
```
Create time (s)
('br', 1): 0.81
('br', 2): 0.93
('br', 3): 1.09
('br', 4): 2.95
('br', 5): 7.21
('br', 6): 12.69
('br', 7): 21.45
(False, None): 0.57
('gzip', None): 6.30

Size (MB)
('br', 1): 133.99
('br', 2): 132.35
('br', 3): 132.25
('br', 4): 132.23
('br', 5): 127.00
('br', 6): 126.25
('br', 7): 126.22
(False, None): 229.82
('gzip', None): 136.68
```

`8bit`
```
Create time (s)
('br', 1): 0.64
('br', 2): 0.65
('br', 3): 0.70
('br', 4): 0.89
('br', 5): 1.49
('br', 6): 2.06
('br', 7): 2.43
(False, None): 0.53
('gzip', None): 1.32

Size (MB)
('br', 1): 93.38
('br', 2): 93.36
('br', 3): 93.20
('br', 4): 92.95
('br', 5): 93.10
('br', 6): 93.11
('br', 7): 93.10
(False, None): 114.91
('gzip', None): 93.61
```

Reply:
I don't really know what to make of this, however.  Seems like cloud-volume uses multiple cores (no matter what setting is specified) and that can make compression benchmarking w/ cloudvolume really difficult.

16 cores, 32 threads

`16bit`
```
Create time (s)
('br', 1): 1.44
('br', 2): 1.37
('br', 3): 1.45
('br', 4): 1.64
('br', 5): 2.16
('br', 6): 3.26
('br', 7): 5.69
(False, None): 1.25
('gzip', None): 3.71
```

`8bit`
```
Create time (s)
('br', 1): 1.26
('br', 2): 1.17
('br', 3): 1.21
('br', 4): 1.21
('br', 5): 1.30
('br', 6): 1.40
('br', 7): 1.46
(False, None): 1.01
('gzip', None): 1.46
```